fem/nonlininteg.hpp

```diff
@@ -44,6 +44,38 @@ class NonlinearFormIntegrator
    virtual ~NonlinearFormIntegrator() { }
 };
 
+class BlockNonlinearFormIntegrator
```

Comment:
I apologize if I just haven't seen this yet, but is there a specific reason we need BlockNonlinearFormIntegrator and cannot just use NonlinearFormIntegrator?

While we're making that change, I think we could also add methods that deal with different test and trial spaces, since they'll be necessary when you have non-square blocks (... I have HDG in mind here).

Reply:
Oh, never mind I spoke too soon. I see the `Array<Vector *>` arguments now. I still think it would be nice to eventually have support for different test and trial spaces, however (second point above). Great work!

Reply:
Yep! The idea here is that one could have terms involving variables from all of the function spaces. In the incompressible elasticity example, this includes terms that involve the product of the pressure and the deformation gradient.

These mixed formulations are possible, but it would involve a new nonlinear form type class that has two vectors of FE spaces: one for the test spaces and another for the trial ones. It shouldn't be too hard to add if this structure is followed.